### PR TITLE
rust-client/: Add rust-libp2p based punchr client

### PR DESCRIPTION
**Pull request by mxinden**
Adds a Punchr client based on rust-libp2p supporting both Circuit Relay v1 and v2 based on a command line flag.

Credit goes to @elenaf9.

### CODE REVIEW

File: punchr.proto [hunk 1]
```diff
@@ -124,9 +124,16 @@ message HolePunchAttempt {
...
+  // TODO: How about making it an enum?
+  //
   // The error that occurred if the hole punch failed
   optional string error = 7;
 
+  // TODO mxinden: I am not quite sure I understand this field. Why does the
+  // server pass a public non-relay address to the client in the first place?
+  // Also note that connection reversal is only attempted by the DCUtR listener,
+  // not dialer.
+  //
```
Comment (mxinden) [inline punchr.proto:136]:
```suggestion
```

File: punchr.proto [hunk 2]
```diff
@@ -124,9 +124,16 @@ message HolePunchAttempt {
   // The outcome of the hole punch
   required HolePunchAttemptOutcome outcome = 6;
 
+  // TODO: How about making it an enum?
+  //
```
Comment (mxinden) [inline punchr.proto:128]:
```suggestion
```

Obsolete with HolePunchAttemptOutcome.

File: punchr.proto [hunk 3]
```diff
@@ -127,6 +127,11 @@ message HolePunchAttempt {
   // The error that occurred if the hole punch failed
   optional string error = 7;
 
+  // TODO mxinden: I am not quite sure I understand this field. Why does the
```
Comment (mxinden) [inline punchr.proto:130]: @dennis-tra I might be missing something. Is this field still needed?

**Comment by mxinden** [comment]
Apart from the question above, this is ready to be merged from my end. @dennis-tra what do you think?

File: punchr.proto (hunk 3 above)
Comment (dennis-tra) [reply punchr.proto:130]: This is just another code path that could in theory be reached from the Go implementation and I just wanted to be able to track all potential information from the hole punch. The most interesting research results come from things that should never happen but do anyways :D I would opt for keeping it and just rephrase the comment to make it unambiguously clear that this field should never have a value.

**Comment by dennis-tra** [comment]
Ready when you are :) I just invited you two to the repository as collaborators so feel free to merge it 👍

**Review by elenaf9 — APPROVED** [review]

**Pull request merged**